docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,438 @@
+	// +optional
+	Values string `json:"values,omitempty"`
+}
+
+// ClusterSelectorLabel defines a key/value pair used to select Clusters with a label matching the specified key and value.
+type ClusterSelectorLabel struct {
+	// Key is the label key.
+	Key string `json:"key"`
+
+	// Value is the label value.
+	Value string `json:"value"`
+}
+
+// HelmChartProxyStatus defines the observed state of HelmChartProxy.
+type HelmChartProxyStatus struct {
+	// Ready is true when the HelmReleaseProxySpec for each selected Cluster is up to date.
+	// +optional
+	Ready bool `json:"ready"`
+
+	// MatchingClusters is the list of references to Clusters selected by the ClusterSelectorLabel.
+	// +optional
+	MatchingClusters []corev1.ObjectReference `json:"matchingClusters"`
+
+	// FailureReason will be set in the event that there is a an error reconciling the HelmChartProxy.
+	// +optional
+	FailureReason string `json:"failureReason,omitempty"`
+}
+```
+
+#### HelmReleaseProxy CRD
+
+The second CRD, HelmReleaseProxy, is responsible for installing and orchestrating the Helm release on a selected workload Cluster. The HelmReleaseProxy is created as part of the parent HelmChartProxy flow. The parent HelmChartProxy will be responsible for keeping its HelmReleaseProxy children up-to-date as its lifecycle is updated (for example, if the `values` properties of a HelmChartProxy are updated, its child HelmReleaseProxy resources will be updated as well).
+
+```go
+// HelmReleaseProxySpec defines the desired state of HelmReleaseProxy.
+type HelmReleaseProxySpec struct {
+	// ClusterRef is a reference to the Cluster to install the Helm release on.
+	ClusterRef *corev1.ObjectReference `json:"clusterRef,omitempty"`
+
+	// ChartName is the name of the Helm chart in the repository.
+	ChartName string `json:"chartName,omitempty"`
+
+	// RepoURL is the URL of the Helm chart repository.
+	RepoURL string `json:"repoURL,omitempty"`
+
+	// ReleaseName is the release name of the installed Helm chart. If it is not specified, a name will be generated.
+	// +optional
+	ReleaseName string `json:"releaseName,omitempty"`
+
+	// Version is the version of the Helm chart. To be replaced with a compatibility matrix.
+	// +optional
+	Version string `json:"version,omitempty"`
+
+	// Namespace is the namespace the Helm release will be installed on the referenced Cluster. If it is not specified, the default namespace.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Values is an inline YAML representing the values for the Helm chart. This YAML is the result of the rendered
+	// Go templating with the values from the referenced workload Cluster.
+	// +optional
+	Values string `json:"values,omitempty"`
+}
+
+// HelmReleaseProxyStatus defines the observed state of HelmReleaseProxy.
+type HelmReleaseProxyStatus struct {
+	// Ready is true when the Helm release on the referenced Cluster is up to date with the HelmReleaseProxySpec.
+	// +optional
+	Ready bool `json:"ready"`
```

Comment:
similar question here about the `bool` usage, should we consider possible alternate states in the future?

Reply:
Similar to my comment above, I'm working on adding a `ClusterAvailableCondition` to signal that we can access the Cluster and get its kubeconfig and a `HelmReleaseReadyCondition` to show that the Helm release was successfully installed. Does that answer your question about adding additional states or do you think we should add something explicit like `Phase` to the status?

Reply:
it does answer my question, i was hoping to see us not use a `bool` here, but it sounds like you've got an alternative idea about getting the deeper status from the conditions. i'm assuming those conditions would be on the `status` field then?

Reply:
I'll just answer both questions here since it's about the same thing in both CRDs. I've added a `clusterv1.Conditions` field below in the status structs like this.
```
// Conditions defines current state of the HelmReleaseProxy.
// +optional
Conditions clusterv1.Conditions `json:"conditions,omitempty"`
```

 I did have a question about the difference between the `Ready` boolean and the `ReadyCondition` which is aggregated using `conditions.SetSummary()`. From what I've seen most CAPI CRD statuses have a `Ready` field and a `FailureReason` and `FailureMessage` field, but it also includes a `ReadyCondition` which itself has its own `Reason` and `Message` fields. Is there a reason to have both, and if so, what's the difference in meaning and usage? If the goal is to deprecate the `Ready` field then maybe we can drop it from the new struct type here.

Reply:
thanks for the explanation, it makes sense and is also confusing XD

i think the questions you are asking are great, i'm curious about those answers too. 

> If the goal is to deprecate the Ready field then maybe we can drop it from the new struct type here.

this seems important, because yeah maybe we can just avoid the whole field. it's not a blocker for me, i think it would be good if we drop the boolean but i think your reasoning for the conditions and the field make sense.

Reply:
I *think* the bool fields are legacy and the conditions the future

(cc @fabriziopandini to be sure)

Reply:
Yep, if we're sure we want to replace the `ready` bool in favor of a summary condition then I can drop it from the status.